### ISSUE TITLE
Documentation for expected import format

### ISSUE THREAD

**Issue by elyobo** [judilsteve/asx-cgt-calculator-optimiser#1]
If you could provide documentation for the expected import format it would make it easier to use the tool, as we can (hopefully) transform transaction data from other systems into the expected import format.

**Comment by judilsteve** [comment]
Hi mate,

I can definitely see how a bulk import for holdings would be useful, however the original intention of the import/export buttons was simply to allow users to back up and restore their holdings, or transfer them between computers.

I have been mulling over a CSV/Excel based system for bulk inserting data from external sources, but can't provide any promises there.

In terms of documenting the original import system, I think the best way to go might be to port the project to typescript so that type annotations are available, and then include a [step](https://www.npmjs.com/package/typescript-json-schema) in the build that transforms the type specifications into a [JSON schema](https://json-schema.org//) and places the schema file into the static content folder of the build output. I want to avoid manually documenting these sorts of things as much as possible, because of the large risk of the documentation going out of sync with what's actually in the code.

In the meantime, the best way to get this information would be to make a single dummy row in each table, then hit the export button. You can run the resulting file through a [formatter](https://jsonformatter.curiousconcept.com/) to make it easier to read. 

For convenience, here's the output of the above process as at commit 0994ad34f94e5d1f89db139b93f409e3c3c07cd8 (the currently deployed version as at 2021-07-15):

```json
{
   "parcels":[
      {
         "id":"A",
         "date":"2021-07-14T23:55:00.000Z",
         "asxCode":"ASXCODE",
         "isDrp":false,
         "memo":"MEMO",
         "units":123,
         "unitPrice":42,
         "brokerage":42
      }
   ],
   "adjustments":[
      {
         "id":"6b925bcb-41db-4e80-895f-37ea3fcbbc43",
         "date":"2021-07-14T23:56:00.000Z",
         "asxCode":"ASXCODE",
         "applicableParcelIds":[
            "A"
         ],
         "memo":"MEMO",
         "netAmount":42
      }
   ],
   "sales":[
      {
         "id":"6ad3cd96-060d-434b-b681-5c7e49f94d7b",
         "date":"2021-07-14T23:56:00.000Z",
         "asxCode":"ASXCODE",
         "applicableParcels":[
            {
               "id":"A",
               "unitsSold":42
            }
         ],
         "memo":"MEMO",
         "unitPrice":42,
         "brokerage":42
      }
   ]
}
```

Note that for each sale/adjustment, you will need to manually generate a [GUID](https://www.guidgenerator.com/) to put in the ID field.

**Comment by elyobo** [comment]
Thanks, appreciate that it's complex, but the above is helpful guidance! Your suggested approach regarding a JSON schema built from the typescript definitions to avoid drift between docs and code makes sense, but I'll just mention that if the code is changing that's going to be problematic for other reasons as well (i.e. old exports are no longer importable if the schema changes in backwards incompatible ways).

**Comment by judilsteve** [comment]
You're absolutely right about schema changes needing special handling. If there are going to be any changes to the schema, I'll add a `schemaVersion` property to the root of the JSON which can be interrogated at page load and/or import time to conditionally apply a sequence of migration operations, much like the EF Core migration system.
